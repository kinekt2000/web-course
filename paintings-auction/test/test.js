const assert = require("assert")
const should = require("should")

const paintings = require("../../Auction2/public/json/paintings.json")
const config = require("../../Auction2/public/json/config.json")
const users = require("../../Auction2/public/json/users.json")

describe("paintings", () => {
    for(let i = 0; i < paintings.length; i++) {
        const painting = paintings[i]

        describe(`painting ${i}`, () => {
            describe(`properties`, () =>{
                it("has property id", () => {
                    painting.should.have.property("id");
                })
                it("has property name", () => {
                    painting.should.have.property("name");
                })
                it("has property author", () => {
                    painting.should.have.property("author");
                })
                it("has property price", () => {
                    painting.should.have.property("price");

                })
                it("has property step", () => {
                    painting.should.have.property("step");

                })
                it("has property description", () => {
                    painting.should.have.property("description");
                })
            })

            describe(`types`, () => {
                it("id is number type", () => {
                    painting.id.should.be.type("number");
                })
                it("name is string type", () => {
                    painting.name.should.be.type("string");
                })
                it("author is string type", () => {
                    painting.author.should.be.type("string");
                })
                it("price is number type", () => {
                    painting.price.should.be.type("number");
                })
                it("step is number type", () => {
                    painting.step.should.be.type("number");
                })
                it("description is string type", () => {
                    painting.description.should.be.type("string");
                })
            })
        })

    }
})

describe("users", () => {
    for(let i = 0; i < users.length; i++) {
        const user = users[i]

        describe(`user ${i}`, () => {
            describe(`properties`, () => {
                it("has property id", () => {
                    user.should.have.property("id")
                })
                it("has property name", () => {
                    user.should.have.property("name")
                })
                it("has property fund", () => {
                    user.should.have.property("fund")
                })
            })

            describe(`types`, () => {
                it("id is number type", () => {
                    user.id.should.be.type("number")
                })
                it("name is string type", () => {
                    user.name.should.be.type("string")
                })
                it("fund is number type", () => {
                    user.fund.should.be.type("number")
                })
            })
        })
    }
})

function timeToInt(time) {
    const [first, second] = time.split(":");
    const iFirst = parseInt(first);
    const iSecond = parseInt(second);
    if(isNaN(iFirst) || isNaN(iSecond)) {
        return null;
    } else {
        return iFirst * 60 + iSecond
    }
}

describe("config", () => {
    describe("date", () => {
        const [y, m, d] = config.auction.date.split("/");
        it("format is 'yy/mm/dd'", () => {
            if(y === undefined) assert.fail("year is undefined")
            if(m === undefined) assert.fail("month is undefined")
            if(d === undefined) assert.fail("day is undefined")

            if(y.length !== 4 || isNaN(parseInt(y))) assert.fail("year has wrong format")
            if(m.length !== 2 || isNaN(parseInt(m))) assert.fail("month has wrong format")
            if(d.length !== 2 || isNaN(parseInt(d))) assert.fail("day has wrong format")
        })

        it("date validation", () => {
            const year = parseInt(y),
                month = parseInt(m),
                day = parseInt(d)

            const date = new Date(year, month - 1, day);
            assert.strictEqual(date.getFullYear(), year)
            assert.strictEqual(date.getMonth() + 1, month)
            assert.strictEqual(date.getDate(), day)

            if(date < new Date()) assert.fail("date must be equal or bigger than today")
        })
    })

    describe("time properties", () => {
        const time = config.auction.time;
        const timeout = config.auction.timeout;
        const countdown = config.auction.countdown;
        const pause = config.auction.pause;

        it("time validation", () => {
            if(isNaN(timeToInt(time))) assert.fail("wrong time")
        })
        it("timeout validation", () => {
            if(isNaN(timeToInt(timeout))) assert.fail("wrong time")
        })
        it("countdown validation", () => {
            if(isNaN(timeToInt(countdown))) assert.fail("wrong time")
        })
        it("pause validation", () => {
            if(isNaN(timeToInt(pause))) assert.fail("wrong time")
        })
    })
})

