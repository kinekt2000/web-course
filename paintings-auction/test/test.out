

  paintings
    properties 0
      ✓ has property id
      ✓ has property name
      ✓ has property author
      ✓ has property price
      ✓ has property step
      ✓ has property description
    types 0
      ✓ id is number type
      ✓ name is string type
      ✓ author is string type
      ✓ price is number type
      ✓ step is number type
      ✓ description is string type
    properties 1
      ✓ has property id
      ✓ has property name
      ✓ has property author
      ✓ has property price
      ✓ has property step
      ✓ has property description
    types 1
      ✓ id is number type
      ✓ name is string type
      ✓ author is string type
      ✓ price is number type
      ✓ step is number type
      ✓ description is string type
    properties 2
      ✓ has property id
      ✓ has property name
      ✓ has property author
      ✓ has property price
      ✓ has property step
      ✓ has property description
    types 2
      ✓ id is number type
      ✓ name is string type
      ✓ author is string type
      ✓ price is number type
      ✓ step is number type
      ✓ description is string type
    properties 3
      ✓ has property id
      ✓ has property name
      ✓ has property author
      ✓ has property price
      ✓ has property step
      ✓ has property description
    types 3
      ✓ id is number type
      ✓ name is string type
      ✓ author is string type
      ✓ price is number type
      ✓ step is number type
      ✓ description is string type

  users
    properties 0
      ✓ has property id
      ✓ has property name
      ✓ has property fund
    types 0
      ✓ id is number type
      ✓ name is string type
      ✓ fund is number type
    properties 1
      ✓ has property id
      ✓ has property name
      ✓ has property fund
    types 1
      ✓ id is number type
      ✓ name is string type
      ✓ fund is number type
    properties 2
      ✓ has property id
      ✓ has property name
      ✓ has property fund
    types 2
      ✓ id is number type
      ✓ name is string type
      ✓ fund is number type
    properties 3
      ✓ has property id
      ✓ has property name
      ✓ has property fund
    types 3
      ✓ id is number type
      ✓ name is string type
      ✓ fund is number type
    properties 4
      ✓ has property id
      ✓ has property name
      ✓ has property fund
    types 4
      ✓ id is number type
      ✓ name is string type
      ✓ fund is number type
    properties 5
      ✓ has property id
      ✓ has property name
      ✓ has property fund
    types 5
      ✓ id is number type
      ✓ name is string type
      ✓ fund is number type
    properties 6
      ✓ has property id
      ✓ has property name
      ✓ has property fund
    types 6
      ✓ id is number type
      ✓ name is string type
      ✓ fund is number type
    properties 7
      ✓ has property id
      ✓ has property name
      ✓ has property fund
    types 7
      ✓ id is number type
      ✓ name is string type
      ✓ fund is number type

  config
    date
      ✓ format is 'yy/mm/dd'
      1) date validation
    time properties
      ✓ time validation
      ✓ timeout validation
      ✓ countdown validation
      ✓ pause validation


  101 passing (25ms)
  1 failing

  1) config
       date
         date validation:

      AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:

12 !== 11

      + expected - actual

      -12
      +11
      
      at Context.<anonymous> (test/test.js:119:20)
      at processImmediate (internal/timers.js:456:21)



